pkg/operator/controller/machineconfig/controller.go

```diff
@@ -80,228 +86,139 @@ func New(mgr ctrl.Manager) *MachineConfigReconciler {
-		r.Log.V(1).Info("node labels were removed, waiting for worker MCP to start updating")
-		return r.waitWorkerMCPStatusUpdating(ctx)
+		return ctrl.Result{}, fmt.Errorf("failed to update nodeobservabilitymachineconfig with finalizers: %w", err)
 	}
 
-	if r.CtrlConfig.Status.IsMachineConfigInProgress() {
-		if result, err := r.checkWorkerMCPStatus(ctx); err != nil {
-			return result, err
-		} else if result.Requeue || result.RequeueAfter != 0 {
-			// requeue was asked which means that
-			// we are waiting for the machine updates,
-			// recheck the status of MCP later
-			r.Log.V(1).Info("waiting for worker MCP to stabilize")
-			return result, nil
-		}
+	if err := r.handleProfilingRequest(ctx, nomc); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile requested configuration: %w", err)
 	}
 
-	removeFinalizer := false
-	if !r.CtrlConfig.Status.IsMachineConfigInProgress() {
-		if !r.CtrlConfig.Status.IsDebuggingEnabled() {
-			removeFinalizer = true
-			r.Log.V(1).Info("disable debug successful for cleanup")
-		}
-
-		if r.CtrlConfig.Status.IsDebuggingFailed() {
-			removeFinalizer = true
-			r.Log.V(1).Info("failed to disable debug for cleanup")
+	// update nomc status based on mcp status conditions
+	// after handling profiling request
+
+	// if nodeobservabilitymachineconfig status is DebugEnabled=True
+	// we poll until we sync nodeobservability mcp
+	// completely
+	if nomc.Status.IsDebuggingEnabled() && nomc.Status.IsMachineConfigInProgress() {
+		// sync for nodeobservability mcp status
+		if requeue, err := r.syncNodeObservabilityMCPStatus(ctx, nomc); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to check nodeobservability mcp status: %w", err)
+		} else if requeue {
+			return ctrl.Result{RequeueAfter: defaultRequeueTime}, nil
 		}
 	}
 
-	if removeFinalizer && hasFinalizer(r.CtrlConfig) {
-		if _, err := r.removeFinalizer(ctx, req, finalizer); err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from nodeobservabilitymachineconfig %s: %w", r.CtrlConfig.Name, err)
+	// if nodeobservabilitymachineconfig status is DebugEnabled=False
+	// we poll until we sync worker mcp completely
+	if !nomc.Status.IsDebuggingEnabled() && nomc.Status.GetCondition(v1alpha2.DebugEnabled).Reason == v1alpha2.ReasonDisabled && nomc.Status.IsMachineConfigInProgress() {
+		// sync for worker mcp status
+		if requeue, err := r.syncWorkerMCPStatus(ctx, nomc); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to check worker mcp status: %w", err)
+		} else if requeue {
+			return ctrl.Result{RequeueAfter: defaultRequeueTime}, nil
 		}
-		r.Log.V(1).Info("removed finalizer from nodeobservabilitymachineconfig resource, cleanup complete")
-		return ctrl.Result{}, nil
 	}
 
-	r.Log.V(1).Info("cleanup is not complete, retry")
-
-	return ctrl.Result{RequeueAfter: defaultRequeueTime}, nil
+	r.Log.V(2).Info("reconciliation completed", "nodeobservabilitymachineconfig", req.NamespacedName)
```

Comment:
```suggestion
	r.Log.V(1).Info("reconciliation completed")
```
We report the start of the reconciliation as V1, let's do the se for the end.